### PR TITLE
Locations lookup

### PR DESCRIPTION
**Pull request by slorek**
This allows the frontend to look up the user's locations either on HMPPS SSO or NOMIS depending on the type of user, and resolves those to the IDs represented in the API. This will allow us to implement location switching.

### CODE REVIEW

File: app/auth/middleware.test.js [hunk 1]
```diff
@@ -1,13 +1,13 @@
...
-  this.user_name = token.user_name
+const locationStub = {
+  getUserLocations: token => {
+    return ['TEST']
+  },
 }
 
 const authentication = proxyquire('./middleware', {
-  '../../common/lib/user': UserStub,
+  '../../common/services/user-locations': locationStub,
```
Comment (teneightfive) [inline app/auth/middleware.test.js]: This looks like we're not longer stubbing the User class. Why has this been removed?

File: common/services/user-locations.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+      break
+  }
+
+  const locations = await Promise.all(
+    agencies.map(async agencyId => {
+      return referenceDataService.getLocations({ nomisAgencyId: agencyId })
+    })
+  )
+
+  return locations.map(results => {
```
Comment (teneightfive) [inline common/services/user-locations.js]: Could be shortened to:

```js
return locations.map(results => results[0].id)
```

File: common/services/user-locations.js [hunk 3]
```diff
@@ -0,0 +1,68 @@
+const axios = require('axios')
+
+const { AUTH_PROVIDERS, NOMIS_ELITE2_API } = require('../../config')
+const { decodeAccessToken } = require('../lib/access-token')
+const referenceDataService = require('./reference-data')
+
+async function getUserLocations(token) {
+  var agencies
```
Comment (teneightfive) [inline common/services/user-locations.js]: Better to favour `let` over `var`

File: common/services/user-locations.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+async function _getAuthAgencies(token) {
+  const tokenData = decodeAccessToken(token)
+
+  const response = await axios.get(
+    AUTH_PROVIDERS.hmpps.groups_url(tokenData.user_name),
+    {},
+    _authHeaders(token)
+  )
+
+  return response.data.map(group => {
```
Comment (teneightfive) [inline common/services/user-locations.js]: Could be shortened to:

```js
return response.data.map(group => group.groupCode.replace(/^PECS_/, ''))
```

File: common/services/user-locations.js [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+async function _getNomisAgencies(token) {
+  const response = await axios.get(
+    NOMIS_ELITE2_API.user_caseloads_url,
+    {},
+    _authHeaders(token)
+  )
+
+  return response.data.map(caseLoad => {
```
Comment (teneightfive) [inline common/services/user-locations.js]: Could be shortened to:

```js
return response.data.map(caseLoad => caseLoad.caseLoadId)
```

or, if you use [lodash's map](https://lodash.com/docs/4.17.15#map) function it could be:

```js
return map(response.data, 'caseLoadId')
```

File: common/services/user-locations.js [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    {},
+    _authHeaders(token)
+  )
+
+  return response.data.map(group => {
+    return group.groupCode.replace(/^PECS_/, '')
+  })
+}
+
+async function _getNomisAgencies(token) {
```
Comment (teneightfive) [inline common/services/user-locations.js]: I wonder if it might be more readable to make these return a promise rather than using async/await.

This method could then become:

```js
function _getNomisAgencies(token) {
  return axios
    .get(NOMIS_ELITE2_API.user_caseloads_url, {
      headers: _getAuthHeaders(token),
    })
    .then(caseLoad => caseLoad.caseLoadId)
}
```

And I think `_authHeaders` should return just the headers object, not the whole config object:

```js
function _getAuthHeaders(token) {
  return {
    Authorization: `Bearer ${token}`,
  }
}
```

Comment (teneightfive) [reply common/services/user-locations.js]: `_getAuthAgencies` can also be structured in a similar way.

File: common/services/user-locations.js [hunk 7]
```diff
@@ -0,0 +1,68 @@
+const axios = require('axios')
+
+const { AUTH_PROVIDERS, NOMIS_ELITE2_API } = require('../../config')
+const { decodeAccessToken } = require('../lib/access-token')
+const referenceDataService = require('./reference-data')
+
+async function getUserLocations(token) {
```
Comment (teneightfive) [inline common/services/user-locations.js]: After looking at the other methods in this module I think this might be neater if its restructured to return a promise as well rather than use async/await.

File: common/services/user-locations.test.js [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+  getLocations: async args => {
+    return [{ id: 'test' }, { id: 'test2' }]
+  },
+}
+
+const userLocationsService = proxyquire('./user-locations', {
+  './reference-data': referenceDataStub,
+})
+
+const { AUTH_PROVIDERS, NOMIS_ELITE2_API } = require('../../config')
```
Comment (teneightfive) [inline common/services/user-locations.test.js]: I think it's better to stub these using `proxyquire` rather than load them into the test file.

This is the problem we have in the other services at the moment and is causing #82. I think it's what is causing the tests to fail on CI too.

File: config/index.js [hunk 9]
```diff
@@ -81,11 +85,19 @@ module.exports = {
       logout_url: _authUrl(
         `/auth/logout?client_id=${AUTH_KEY}&redirect_uri=${BASE_URL}`
       ),
+      groups_url: userName => {
+        return _authUrl(`/auth/api/authuser/${userName}/groups`)
+      },
+      base_url: AUTH_BASE_URL,
```
Comment (teneightfive) [inline config/index.js]: Do we need to add this? It looks like it's being used by the tests and pulled in direct to create the mock?

I think if we stub the url in the test using `proxyquire` it prevents us needing to add this to the config.

File: config/index.js [hunk 10]
```diff
@@ -81,11 +85,19 @@ module.exports = {
...
       key: AUTH_KEY,
       secret: process.env.AUTH_PROVIDER_SECRET,
     },
   },
   DEFAULT_AUTH_PROVIDER: 'hmpps',
+  NOMIS_ELITE2_API: {
+    user_caseloads_url: _nomisElite2Url('/elite2api/api/users/me/caseLoads'),
+    base_url: NOMIS_ELITE2_API_URL,
```
Comment (teneightfive) [inline config/index.js]: Same as the other config `base_url` config.

File: app/auth/middleware.test.js (hunk 1 above)
Comment (slorek) [reply app/auth/middleware.test.js]: It's no longer necessary since the `User` class is now much simpler and has no external dependencies.

File: common/services/user-locations.js (hunk 6 above)
Comment (slorek) [reply common/services/user-locations.js]: Thanks, I forgot about `then`. I couldn't get the fancy `map` syntax above to work but it's certainly better now.

File: common/services/user-locations.js (hunk 7 above)
Comment (slorek) [reply common/services/user-locations.js]: 🤔 I've refactored this as you suggest and it's cleaned it up a bit, though it leaves a somewhat nasty nested `then` which I think makes it a bit less clear what the return value of the function is. Have a look and let me know what you think.

File: common/services/user-locations.test.js (hunk 8 above)
Comment (slorek) [reply common/services/user-locations.test.js]: 👍

File: config/index.js (hunk 9 above)
Comment (slorek) [reply config/index.js]: Good shout.

File: app/auth/middleware.test.js (hunk 1 above)
Comment (teneightfive) [reply app/auth/middleware.test.js]: I'm a little uncomfortable removing this stub as if the User library changes in the future and gets more complex it will be being pulled into this test and could affect it.

I think it's good to stub services to what you expect in the tests rather than allow it to use all the dependencies.

File: common/services/user-locations.js [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+      return response.data.map(group => group.groupCode.replace(/^PECS_/, ''))
+    })
+}
+
+function _getNomisAgencies(token) {
+  return axios
+    .get(NOMIS_ELITE2_API.user_caseloads_url, {
+      headers: _getAuthHeaders(token),
+    })
+    .then(response => {
```
Comment (teneightfive) [inline common/services/user-locations.js]: Does using lodash map not work here?

Could this work:

```js
.then(response => map(response.data, 'caseLoadId')
```

or maybe even chaining thens to make it easier to read:

```js
.then(response => response.data)
.then(data => map(data, 'caseLoadId'))
```

File: common/services/user-locations.js [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+function _getAuthAgencies(token) {
+  const tokenData = decodeAccessToken(token)
+
+  return axios
+    .get(AUTH_PROVIDERS.hmpps.groups_url(tokenData.user_name), {
+      headers: _getAuthHeaders(token),
+    })
+    .then(response => {
```
Comment (teneightfive) [inline common/services/user-locations.js]: Could do same as other method:

```js
.then(response => response.data)
.then(data => data.map(group => group.groupCode.replace(/^PECS_/, '')))
```

File: common/services/user-locations.js [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+function _getAgencies(token) {
+  let agenciesPromise
+
+  const tokenData = decodeAccessToken(token)
+
+  switch (tokenData.auth_source) {
+    case 'auth':
+      agenciesPromise = _getAuthAgencies(token)
```
Comment (teneightfive) [inline common/services/user-locations.js]: Could just return within the switch to avoid needed to declare the promise:

```js
function _getAgencies(token) {
  const { auth_source: source } = decodeAccessToken(token)

  switch (source) {
    case 'auth':
      return _getAuthAgencies(token)
      break
    case 'nomis':
      return _getNomisAgencies(token)
      break
  }
}
```

Also could destructure the token.

File: common/services/user-locations.js [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+
+const { AUTH_PROVIDERS, NOMIS_ELITE2_API } = require('../../config')
+const { decodeAccessToken } = require('../lib/access-token')
+const referenceDataService = require('./reference-data')
+
+function getUserLocations(token) {
+  return _getAgencies(token).then(agencies => {
+    return Promise.all(
+      agencies.map(agencyId =>
+        referenceDataService.getLocations({ nomisAgencyId: agencyId })
```
Comment (teneightfive) [inline common/services/user-locations.js]: Just looking at this, wonder if this should be in the reference data service as `getLocationByAgencyCode` rather than part of the all locations method. It already contains `getLocationById`.

That method could then be responsible for returning the first result if there are multiple by that agency. Removes that responsibility from this library as it shouldn't care. Might make it easier to test too.

Comment (teneightfive) [reply common/services/user-locations.js]: In fact, maybe create a `getLocationsByAgencyId` which mimics [`getLocationsById`](https://github.com/ministryofjustice/hmpps-book-secure-move-frontend/blob/master/common/services/reference-data.js#L52-L60) and call that from this library?

Comment (slorek) [reply common/services/user-locations.js]: This was my original approach but after our in-person discussion I think we have a way forward.

File: common/services/user-locations.js (hunk 13 above)
Comment (teneightfive) [reply common/services/user-locations.js]: Also just thinking about fallback, maybe we should have a `default` case which returns a successful promise incase the source doesn't exist and it doesn't cause an unhandled promise error.

File: app/auth/middleware.js [hunk 15]
```diff
@@ -15,20 +11,16 @@ function processAuthResponse(defaultLocations = []) {
...
           return next(error)
         }
 
-        req.session.authExpiry = accessToken.exp
-        req.session.postAuthRedirect = postAuthRedirect
-        req.session.user = new User({
-          ...accessToken,
-          locations,
-        })
+        _setSessionData(
```
Comment (teneightfive) [inline app/auth/middleware.js]: I'd probably just scrap this. Is it just to beat code climate's warning of line length? If it's just that we can manually ignore that warning for this. I don't think splitting this out helps with re-use or readability to be honest.

If you were to refactor at all it would be neater to separate out the `session.regenerate` callback rather than have half the function inline and the other half separated.

File: common/lib/user.js [hunk 16]
```diff
@@ -1,14 +1,14 @@
-function User(token = {}) {
-  this.userName = token.user_name
-  this.locations = token.locations || []
-  this.permissions = this.getPermissions(token.authorities)
+function User({ name, roles = [], locations = [] } = {}) {
```
Comment (teneightfive) [inline common/lib/user.js]: This is much nicer 👍

File: common/services/reference-data.js [hunk 17]
```diff
@@ -49,14 +49,23 @@ function getLocationById(id) {
...
-  )
+function getLocationsByAgencyId(ids = []) {
+  return _mapLocationIdsToLocations(ids, getLocationByAgencyId)
+}
+
+function _mapLocationIdsToLocations(ids, callback) {
+  const locationPromises = ids.map(id => {
+    return callback(id).catch(() => false)
+  })
+  return Promise.all(locationPromises).then(result => result.filter(Boolean))
```
Comment (teneightfive) [inline common/services/reference-data.js]: Maybe leave `result` as `locations`. Feels more clear to what is returned from those promises.

File: common/services/reference-data.js [hunk 18]
```diff
@@ -49,14 +49,23 @@ function getLocationById(id) {
   return apiClient.find('location', id).then(response => response.data)
 }
 
-function getLocationsById(locations = []) {
-  const locationPromises = locations.map(locationId => {
-    return getLocationById(locationId).catch(() => false)
-  })
+function getLocationByAgencyId(agencyId) {
```
Comment (teneightfive) [inline common/services/reference-data.js]: Could also name the argument `nomisAgencyId` so you can use shorthand in the `getLocations` call.

File: common/services/user-locations.test.js [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+  getLocationsByAgencyId: sinon.spy(args => {
+    return [{ id: 'test' }]
+  }),
+}
+
+const configStub = {
+  AUTH_PROVIDERS: {
+    hmpps: {
+      groups_url: userName => {
+        return 'http://test/'
```
Comment (teneightfive) [inline common/services/user-locations.test.js]: Could shorten this to:

```js
groups_url: () => 'http://test/',
```

File: app/auth/middleware.js [hunk 20]
```diff
@@ -38,6 +30,18 @@ function processAuthResponse(defaultLocations = []) {
...
 
+function _setSessionData(session, token, postAuthRedirect) {
+  const tokenData = decodeAccessToken(token)
+
+  session.authExpiry = tokenData.exp
+  session.postAuthRedirect = postAuthRedirect
+  session.user = new User({
+    name: tokenData.user_name,
+    roles: tokenData.authorities,
+    locations: userLocationService.getUserLocations(token),
```
Comment (teneightfive) [inline app/auth/middleware.js]: Does this not require `await` as the call is async?

Probably neater to leave the locations call on line 19, just inside the `try/catch`, like before. And can then leave the line above the same as it was.

File: app/auth/middleware.js (hunk 15 above)
Comment (slorek) [reply app/auth/middleware.js]: I agree it's not any better. I'm not sure it's possible to extract the callback function easily because it'll need scope to access `req` and `next` which are scoped to the parent `middleware` function. Sure it's possible but it's going to be a meaty refactoring (unless you can think of an easy way). Are you happy for me to revert it back and mark it as won't fix?

File: common/services/reference-data.js (hunk 18 above)
Comment (slorek) [reply common/services/reference-data.js]: I was thinking of renaming the function `getLocationByNomisAgencyId` (and everywhere else it is called `agencyId` to be clearer, do you think this is worthwhile?

File: app/auth/middleware.js (hunk 15 above)
Comment (teneightfive) [reply app/auth/middleware.js]: Yeah, given that would be a little messier too I'd just revert to what it was before and we can mark code climate to ignore the line length in this particular instance.

File: common/services/reference-data.js (hunk 18 above)
Comment (teneightfive) [reply common/services/reference-data.js]: I was also thinking the same thing. I think it's clearer to what it's referring to as these are tied directly to NOMIS.

File: app/auth/middleware.js (hunk 20 above)
Comment (teneightfive) [reply app/auth/middleware.js]: Ah, the test is passing because you're not returning a promise, which differs from what `getUserLocations` actually returns.

If you look how the test worked for the reference data stub you should be able to copy that to test promise rejection and success.

Comment (slorek) [reply app/auth/middleware.js]: I've fixed the test and added test cases for failure.

File: app/auth/middleware.test.js [hunk 21]
```diff
@@ -46,8 +43,13 @@ describe('Authentication middleware', function() {
     })
 
     context('when there is no grant response', function() {
-      beforeEach(function() {
-        authentication.processAuthResponse()(req, {}, nextSpy)
+      beforeEach(async function() {
+        const authentication = proxyquire('./middleware', {
```
Comment (teneightfive) [inline app/auth/middleware.test.js]: You can leave the proxyquire definition at the top this where it was and pass a `sinon.stub()` into the `getUserLocations` method. Then use `. rejects()` and `.resolves()` in the test before functions.

At top of file:
```js
const getUserLocationsStub = sinon.stub()
const authentication = proxyquire('./middleware', {
  '../../common/lib/user': UserStub,
  '../../common/services/user-locations': {
    getUserLocations: getUserLocationsStub,
  },
})
```

In success `beforeEach`:

```js
getUserLocationsStub.resolves(['TEST'])
```

In failure `beforeEach`:

```js
getUserLocationsStub.rejects(userLocationFailureError)
```

Comment (slorek) [reply app/auth/middleware.test.js]: Ah, clever thanks. I took this implementation from another test.

File: .env.example [hunk 22]
```diff
@@ -8,6 +8,7 @@ AUTH_PROVIDER_KEY=
 AUTH_PROVIDER_SECRET=
 AUTH_PROVIDER_URL=
 SERVER_HOST=localhost:3000
+NOMIS_ELITE2_API_URL=https://gateway.t3.nomis-api.hmpps.dsd.io
```
[the remark this answers is not in the document]
Comment (slorek) [reply .env.example]: It’s not considered secret so I think it is helpful to include it here.

[the remark this answers is not in the document]
Comment (teneightfive) [reply .env.example]: It doesn't match what we have at the moment is the only thing. I'd say we should either add all of the non-secret URLs if we think it's helpful, or none if we don't want to expose them. 

Bit odd having half a convention.

[the remark this answers is not in the document]
Comment (slorek) [reply .env.example]: I think we should add the missing ones, but in a separate PR.

File: app/auth/middleware.js [hunk 23]
```diff
@@ -15,19 +11,23 @@ function processAuthResponse(defaultLocations = []) {
...
+      const userLocations = await userLocationService.getUserLocations(
+        grant.response.access_token
+      )
 
       req.session.regenerate(error => {
         if (error) {
           return next(error)
         }
 
+        const accessToken = decodeAccessToken(grant.response.access_token)
```
[the remark this answers is not in the document]
Comment (slorek) [reply app/auth/middleware.js]: I felt that it was a minor improvement to move it as `accessToken` is only used within the scope of the callback function (same reason `userLocations` was moved).

File: app/auth/middleware.test.js (hunk 21 above)
Comment (slorek) [reply app/auth/middleware.test.js]: I've had a go at this, but I don't think the dynamic changing of the stub works in this way. It seems that the behaviour is fixed once `proxyquire` loads it. I think that's why it was also implemented this way in https://github.com/ministryofjustice/hmpps-book-secure-move-frontend/blob/master/app/healthcheck/middleware.test.js. Happy for you to have a crack at it.

**Comment by teneightfive** [comment]
Realised with this PR we're introducing another dependency for the frontend app. We should add a ping endpoint to the heathcheck dependencies.

Happy for this to be part of this PR as this is the PR that introduces the dependency but I'd make it another commit to be clearer.

**Comment by slorek** [comment]
Good spot about the health check. I've added this.
